pkg/image/registry/imagestream/etcd/image.go

```diff
@@ -0,0 +1,222 @@
+				return nil, err
+			}
+			return watch.Filter(w, func(in watch.Event) (out watch.Event, keep bool) {
+				if in.Object == nil {
+					return in, true
+				}
+				// reduce each object to the minimal subset we need for the cache
+				image, ok := in.Object.(*imageapi.Image)
+				if !ok {
+					return in, true
+				}
+				in.Object = &ImageLayers{
+					Name:            image.Name,
+					ResourceVersion: image.ResourceVersion,
+					Layers:          image.DockerImageLayers,
+					Manifest:        manifestFromImage(image),
+				}
+				return in, true
+			}), nil
+		},
+	}, &ImageLayers{}, 0, cache.Indexers{
+		// layers allows fast access to the images with a given layer
+		"layers": func(obj interface{}) ([]string, error) {
+			entry, ok := obj.(*ImageLayers)
+			if !ok {
+				return nil, fmt.Errorf("unexpected cache object %T", obj)
+			}
+			keys := make([]string, 0, len(entry.Layers))
+			for _, layer := range entry.Layers {
+				keys = append(keys, layer.Name)
+			}
+			return keys, nil
+		},
+	})
+	return imageLayerIndex{informer: informer}
+}
+
+// manifestFromImage attempts to find a manifest blob description from
+// an image. Images older than schema2 in Docker do not have a manifest blob.
+func manifestFromImage(image *imageapi.Image) *imageapi.ImageLayer {
+	if image.DockerImageManifestMediaType != "application/vnd.docker.distribution.manifest.v2+json" {
+		return nil
+	}
+	return &imageapi.ImageLayer{
+		Name:      image.DockerImageMetadata.ID,
+		MediaType: image.DockerImageManifestMediaType,
+	}
+}
+
+// ImageLayers is the minimal set of data we need to retain to provide the cache.
+// Unlike a more general informer cache, we do not retain the full object because of
+// the potential size of the objects being stored. Even a small cluster may have 20k
+// or more images in active use.
+type ImageLayers struct {
+	Name            string
+	ResourceVersion string
+	Manifest        *imageapi.ImageLayer
+	Layers          []imageapi.ImageLayer
+}
+
+var (
+	_ runtime.Object = &ImageLayers{}
+	_ metav1.Object  = &ImageLayers{}
+)
+
+func (l *ImageLayers) GetObjectKind() schema.ObjectKind { return &metav1.TypeMeta{} }
+func (l *ImageLayers) DeepCopyObject() runtime.Object {
+	var layers []imageapi.ImageLayer
+	if l.Layers != nil {
+		layers = make([]imageapi.ImageLayer, len(l.Layers))
+		copy(layers, l.Layers)
+	}
+	var manifest *imageapi.ImageLayer
+	if l.Manifest != nil {
+		copied := *l.Manifest
+		manifest = &copied
+	}
+	return &ImageLayers{
+		Name:            l.Name,
+		ResourceVersion: l.ResourceVersion,
+		Manifest:        manifest,
+		Layers:          layers,
+	}
+}
+
+// client-go/cache.SharedIndexInformer hardcodes the key function to assume ObjectMeta.
+// Here we implement the relevant accessors to allow a minimal index to be created.
+// SharedIndexInformer will be refactored to require a more minimal subset of actions
+// in the near future.
+
+func (l *ImageLayers) GetName() string                   { return l.Name }
```

Comment:
ah, hence your complaint.  I wouldn't object to using a full objectmeta in the type to avoid this....